src/github.com/getlantern/detour/ioloop.go

```diff
@@ -0,0 +1,348 @@
+					// select on dc.chClose to avoid blocking
+					// on a channel with no receiver.  same
+					// hereafter.
+					select {
+					case <-dc.chClose:
+					case chRemoveConn <- c:
+					}
+				}
+			}()
+
+		case req := <-dc.chWriteRequest:
+			if atomic.LoadUint32(&dc.expectedConns) == 1 {
+				c := conns.Next()
+				go func() {
+					n, err := c.Write(req.buf)
+					log.Tracef("Wrote %d bytes via %s connection to %s, err: %v", n, c.Type(), dc.addr, err)
+					req.chResult <- ioResult{n, err}
+				}()
+				continue
+			}
+			if !dc.anyDataReceived() {
+				if isNonidempotentHTTPRequest(req.buf) {
+					nonidempotentHTTPRequest = true
+				} else {
+					dc.muWriteBuffer.Lock()
+					_, _ = dc.writeBuffer.Write(req.buf)
+					dc.muWriteBuffer.Unlock()
+				}
+			}
+			var lastN int
+			var lastError error
+			conns.Foreach(func(c conn) bool {
+				if n, err := c.Write(req.buf); err != nil {
+					log.Debugf("Error write to %s connection to %s", c.Type(), dc.addr)
+					dc.closeAndDecrease(c)
+					// intentionally not return c to chConns
+					return false
+				} else {
+					log.Tracef("Wrote %v bytes to %s connection to %s", n, c.Type(), dc.addr)
+					lastN = n
+					return true
+				}
+			})
+			if lastN > 0 {
+				req.chResult <- ioResult{lastN, nil}
+			} else {
+				// simply return last error so caller can have a sense of what happening
+				req.chResult <- ioResult{0, lastError}
+			}
+
+		case req := <-dc.chGetAddr:
+			if conns.Len() == 0 {
+				panic("should have at least one valid connection")
+			}
+			c := conns.Next()
+			if req.isLocal {
+				req.chResult <- c.LocalAddr()
+			} else {
+				req.chResult <- c.RemoteAddr()
+			}
+
+		case <-dc.chClose:
+			conns.Foreach(func(c conn) bool {
+				closeConn(c)
+				return false
+			})
+			return
+		}
+	}
+}
+
+func (dc *Conn) replay(c conn) bool {
+	dc.muWriteBuffer.RLock()
+	defer dc.muWriteBuffer.RUnlock()
+	numBytes := dc.writeBuffer.Len()
+	if numBytes == 0 {
+		return false
+	}
+	log.Tracef("Replay %d previous bytes to %s connection to %s", numBytes, c.Type(), dc.addr)
+	if _, err := c.Write(dc.writeBuffer.Bytes()); err != nil {
+		log.Debugf("Fail to replay %s bytes to %s: %s", numBytes, dc.addr, err)
+		return false
+	}
+	return true
+}
+
+func (dc *Conn) anyDataReceived() bool {
+	return atomic.LoadUint64(&dc.readBytes) > 0
+}
+
+func (dc *Conn) incReadBytes(n int) {
+	atomic.AddUint64(&dc.readBytes, uint64(n))
+}
+func (dc *Conn) closeAndDecrease(c conn) {
+	closeConn(c)
+	atomic.AddUint32(&dc.expectedConns, ^uint32(0))
+}
+
+// close with trace
+func closeConn(c conn) {
+	if err := c.Close(); err != nil {
+		log.Tracef("Error close %s connection to %s: %s", c.Type(), c.RemoteAddr().String(), err)
+	}
+}
+
+var nonidempotentMethods = [][]byte{
+	[]byte("PUT "),
+	[]byte("POST "),
+	[]byte("PATCH "),
+}
+
+// Ref https://tools.ietf.org/html/rfc2616#section-9.1.2
+// We consider the https handshake phase to be idemponent.
+func isNonidempotentHTTPRequest(b []byte) bool {
```

Comment:
This is scary for multiple reasons.

1. What about HTTPS where the verb is encrypted?  This method will identify it as idempotent even though it may not be.

2. What if an application author is misusing an idempotent method like GET to actually alter server-side state?